src/lib.rs

```diff
@@ -36,6 +36,279 @@
+pub use psa_crypto_binding::psa_key_handle_t;
+pub use psa_crypto_binding::psa_key_id_t;
+pub use psa_crypto_binding::psa_key_lifetime_t;
+pub use psa_crypto_binding::psa_key_type_t;
+pub use psa_crypto_binding::psa_key_usage_t;
+pub use psa_crypto_binding::psa_status_t;
+
+// Wrapped types:
+
+#[allow(non_camel_case_types)]
+pub struct psa_key_attributes_t {
+    x: psa_crypto_binding::psa_key_attributes_t,
+}
+
+impl Drop for psa_key_attributes_t {
+    fn drop(&mut self) {
+        wrap_any!(psa_crypto_binding::psa_reset_key_attributes(&mut self.x));
+    }
+}
+
+// Wrapped linkable functions:
+
+pub fn psa_asymmetric_sign(
+    handle: psa_key_handle_t,
+    alg: psa_algorithm_t,
+    hash: *const u8,
+    hash_length: usize,
+    signature: *mut u8,
+    signature_size: usize,
+    signature_length: *mut usize,
+) -> psa_status_t {
+    wrap_status!(psa_crypto_binding::psa_asymmetric_sign(
+        handle,
+        alg,
+        hash,
+        hash_length,
+        signature,
+        signature_size,
+        signature_length
+    ))
+}
+
+pub fn psa_asymmetric_verify(
+    handle: psa_key_handle_t,
+    alg: psa_algorithm_t,
+    hash: *const u8,
+    hash_length: usize,
+    signature: *const u8,
+    signature_length: usize,
+) -> psa_status_t {
+    wrap_status!(psa_crypto_binding::psa_asymmetric_verify(
+        handle,
+        alg,
+        hash,
+        hash_length,
+        signature,
+        signature_length
+    ))
+}
+
+pub fn psa_close_key(handle: psa_key_handle_t) -> psa_status_t {
+    wrap_status!(psa_crypto_binding::psa_close_key(handle))
+}
+
+pub fn psa_destroy_key(handle: psa_key_handle_t) -> psa_status_t {
+    wrap_status!(key_lock!(psa_crypto_binding::psa_destroy_key(handle)))
+}
+
+pub fn psa_export_public_key(
+    handle: psa_key_handle_t,
+    data: *mut u8,
+    data_size: usize,
+    data_length: *mut usize,
+) -> psa_status_t {
+    wrap_status!(psa_crypto_binding::psa_export_public_key(
+        handle,
+        data,
+        data_size,
+        data_length
+    ))
+}
+
+pub fn psa_generate_key(
+    attributes: *const psa_key_attributes_t,
+    handle: *mut psa_key_handle_t,
+) -> psa_status_t {
+    wrap_status!(key_lock!(psa_crypto_binding::psa_generate_key(
+        &(*attributes).x,
+        handle
+    )))
+}
+
+pub fn psa_get_key_attributes(
+    handle: psa_key_handle_t,
+    attributes: *mut psa_key_attributes_t,
+) -> psa_status_t {
+    wrap_status!(psa_crypto_binding::psa_get_key_attributes(
+        handle,
+        &mut (*attributes).x
+    ))
+}
+
+pub fn psa_get_key_bits(attributes: &psa_key_attributes_t) -> usize {
+    wrap_any!((*attributes).x.core.bits as usize)
+}
+
+pub fn psa_get_key_type(attributes: &psa_key_attributes_t) -> psa_key_type_t {
+    wrap_any!((*attributes).x.core.type_)
+}
+
+pub fn psa_import_key(
+    attributes: *const psa_key_attributes_t,
+    data: *const u8,
+    data_length: usize,
+    handle: *mut psa_key_handle_t,
+) -> psa_status_t {
```

Comment:
_ignore_

I think we ought to aim to not have any unsafe/non-idiomatic types in our public API. In particular:

* a Rust abstraction for `psa_key_attributes_t` and `psa_algorithm_t`. The setters/getters on the key attributes could even be implemented as methods on the new type we create. See #2 
* in our APIs, instead of having for buffer a raw pointer (`*const u8`) and a size, have a `Vec<u8>` and use methods on it to convert it to what the C API expects. See #3 
* have all our operations return a `Result`, which contains the data that the operation returns (or `()` if nothing) or an error (a `psa_status_t`, abstracted) if there was an error

Reply:
All worth discussing. Currently I have wrapped `psa_key_attributes_t`, so its internals are hidden, I think, but I haven't wrapped `psa_algorithm_t`, which is therefore visible to the user of the Rust wrapper as a `u32` (with the current Mbed implementation).